src/main/scala/com/advancedtelematic/campaigner/http/CampaignResource.scala

```diff
@@ -55,6 +61,12 @@ class CampaignResource(extractAuth: Directive1[AuthedNamespaceScope], director:
       onSuccess(campaigns.findNamespaceCampaign(namespace, campaign)).flatMap(provide)
     }
 
+  private def fetchDevicesInGroups(ns: Namespace, groups: NonEmptyList[GroupId]): Future[Set[DeviceId]] = {
+    val groupSet = groups.toList.toSet
+    // TODO review the retrieval logic and 'limit' parameter
+    Future.sequence(groupSet.map(gid => deviceRegistry.devicesInGroup(ns, gid, 0, 50))).map(_.flatten)
```

Comment:
I think this is a good use case for Future.traverse:
`Future.traverse(groupSet)(gid => deviceRegistry.devicesInGroup(ns, gid, 0, 50)).map(_.flatten)`.